### PR TITLE
Fix the metrics issue in a nested trainer

### LINKED ISSUE #20188: Cannot get result() since the metric has not yet been built
Basically the heading. I am attaching the colab notebook for reproducibility purpose, and have printed out the tensorflow and keras versions for perusal. [Link here](https://colab.research.google.com/drive/1F5Wzq6AqbYTzlBZHXypgiTn9xUVMOhyt#scrollTo=PUKQ6_Vv3nGk)

I think I'm doing everything the standard way. Cannot figure out why this error occurs. Of course this is a simpler code snippet than the one I'm working on which cannot be shared for obvious purposes, but this code demonstrates the issue.

Any clue on how to fix this/circumvent, or which version of keras/tf can i switch to, to avoid this issue?

If it is any help, I saw another [issue](https://github.com/keras-team/keras-nlp/issues/1716) open on keras-nlp, which discusses the same problem, so i'm guessing it's an issue with CompileMetrics module in the trainers folder, which is where the error points anyway.

### PR DESCRIPTION
**Pull request by james77777778**
Fix #20188

The root cause is that we might have unbuilt (and unusable) metrics with a nested trainer.
The solution is to clear these metircs and skip them when calling `self.metrics` in the newly instantiated trainer.

### CODE REVIEW

**Review by fchollet — APPROVED** [review]
Thanks for the fix

**Comment by mattdangerw** [comment]
@james77777778 looks like this broke KerasNLP on nightly.

https://github.com/keras-team/keras-nlp/actions/runs/10687238938/job/29624381773

Do we need to check with `hasattr` or `getattr` somewhere in here? Or is KerasNLP doing something wrong?

**Comment by james77777778** [comment]
I just noticed this. Will try to fix it ASAP.

Using hasattr might be the solution.
